### PR TITLE
SKU reviews dynamic variation

### PR DESCRIPTION
**Pull request by alexunder193**

### CODE REVIEW

File: readme.txt [hunk 1]
```diff
@@ -84,16 +88,7 @@ The following rules apply:
 * If a product is a variant and has two of the specified grouping attributes the `<parent product ID>-<attribute1 termA ID>-<attribute2 termB ID>` is used.
 
 = None of the settings for the Unique product ID works for me. Can I have a custom Unique ID? =
```
Comment (mrwhizzy) [inline readme.txt]: [minor] typo
```suggestion
= None of the settings for the Unique product ID work for me. Can I have a custom Unique ID? =
```

File: includes/class-wc-skroutz-analytics-product.php [hunk 2]
```diff
@@ -77,6 +77,39 @@ public function get_id() {
 		return apply_filters( 'wc_skroutz_analytics_product_id_filter', $product_id, $this->product );
 	}
 
+	/**
+	 * Returns the product variation id that should be reported to Analytics based on
+	 * product id admin settings.
+	 *
+	 * @return string|integer  The product variation id that should be reported to Analytics
+	 *
+	 * @since    1.4.0
```
Comment (MrGoumX) [inline includes/class-wc-skroutz-analytics-product.php]: [minor] We need to change the version here.

File: includes/class-wc-skroutz-analytics-product.php [hunk 3]
```diff
@@ -146,6 +179,34 @@ private function get_product_terms_suffix() {
...
+		return $terms;
+	}
+
+		/**
+	 * Get the suffix of the product id based on the grouped attributes settings.
+	 * Format: -attribute_1_term_id_1-attribute_2_term_id_1...
+	 *
+	 * @return NULL|string The product id suffix. If no product attribute matches any attribute in settings returns null.
+	 *
+	 * @since    1.5.0
```
Comment (MrGoumX) [inline includes/class-wc-skroutz-analytics-product.php]: [minor] Same here.

File: includes/class-wc-skroutz-analytics-product.php [hunk 4]
```diff
@@ -146,6 +179,34 @@ private function get_product_terms_suffix() {
 			}
 		}
 
+		return $terms;
+	}
+
+		/**
```
Comment (MrGoumX) [inline includes/class-wc-skroutz-analytics-product.php]: [ultra-minor] Styling

File: README.md [hunk 5]
```diff
@@ -26,6 +26,11 @@ The plugin is available from the [Wordpress plugin repo][11].
 3. Activate the plugin through the Plugins menu in WordPress admin.
 4. Set the `Shop Account ID` to the plugin's settings.
 
+#### Filters
+
+1. [wc_skroutz_analytics_product_id_filter][13]
+2. [wc_skroutz_analytics_product_variation_id_filter][14]
```
Comment (MrGoumX) [inline README.md]: I think we can change these to anchor links

```suggestion
1. [wc_skroutz_analytics_product_id_filter](#i-dont-use-neither-product-id-nor-sku-in-xml-but-a-custom-postmeta-id)
2. [wc_skroutz_analytics_product_variation_id_filter](#after-integrating-product-reviews-service-i-cannot-see-reviews-for-variation-products)
```

File: README.md [hunk 6]
```diff
@@ -133,4 +150,6 @@ Skroutz Analytics WooCommerce Plugin is licensed under the [GNU Public License v
 [10]: https://merchants.skroutz.gr/merchants/account/settings/analytics
 [11]: https://wordpress.org/plugins/skroutz-analytics-woocommerce/
 [12]: https://developer.wordpress.org/plugins/hooks/filters/
+[13]: https://github.com/skroutz/skroutz-analytics-woocommerce#none-of-the-settings-for-the-unique-product-id-works-for-me-can-i-have-a-custom-unique-id
+[14]: https://github.com/skroutz/skroutz-analytics-woocommerce#after-integrating-product-reviews-service-i-cannot-see-reviews-for-variation-products
```
Comment (MrGoumX) [inline README.md]: And then drop these

**Review by MrGoumX — APPROVED** [review]
+1 LGTM! :rocket:

File: includes/class-wc-skroutz-analytics-product.php [hunk 7]
```diff
@@ -77,6 +77,39 @@ public function get_id() {
...
 
+	/**
+	 * Returns the product variation id that should be reported to Analytics based on
+	 * product id admin settings.
+	 *
+	 * @return string|integer  The product variation id that should be reported to Analytics
+	 *
+	 * @since    1.7.0
+	 */
+	public function get_variable_id() {
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: I believe we should rename this function as the current name may be misleading. This id is tailor made to the product reviews widgets, as it does a best effort to guess the proper product id found in the XML. Maybe rename it to something like: `get_reviews_product_id` or `get_reviews_widget_product_id`.

Also we should update the function comments, to better describe the purpose of this function. This id is not expected to be reported to Analytics.

File: includes/class-wc-skroutz-analytics-product.php [hunk 8]
```diff
@@ -77,6 +77,39 @@ public function get_id() {
...
+		if ( ! $product_id ) {
+			// If product sku is not found set a default id. Analytics does not accept line items without a product id.
+			$product_id = "wc-sa-{$this->product->get_id()}";
+		}
+
+		if ( $this->product->is_type( 'variable' ) && $this->items_product_id_settings['parent_id_enabled'] == 'parent_id_term_id' ) {
+			$product_id .= $this->get_product_first_term_suffix();
+		}
+
+		return apply_filters( 'wc_skroutz_analytics_product_variation_id_filter', $product_id, $this->product );
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: The filter name should also be changed based on the renamed function name

File: includes/class-wc-skroutz-analytics-product.php [hunk 9]
```diff
@@ -77,6 +77,39 @@ public function get_id() {
...
+		}
+
+		if ( ! $product_id ) {
+			$product_id = $this->items_product_id_settings['id'] == 'sku'
+				? $this->product->get_sku()
+				: $this->product->get_id();
+		}
+
+		if ( ! $product_id ) {
+			// If product sku is not found set a default id. Analytics does not accept line items without a product id.
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: We should ✂️ this comment

File: includes/class-wc-skroutz-analytics-product.php [hunk 10]
```diff
@@ -148,4 +181,32 @@ private function get_product_terms_suffix() {
...
+	/**
+	 * Get the suffix of the product id based on the grouped attributes settings.
+	 * Format: -attribute_1_term_id_1-attribute_2_term_id_1...
+	 *
+	 * @return NULL|string The product id suffix. If no product attribute matches any attribute in settings returns null.
+	 *
+	 * @since    1.7.0
+	 * @access   private
+	 */
+	private function get_product_first_term_suffix() {
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php:201]: We should update the function comments, to better describe what the purpose of this function.

File: includes/class-wc-skroutz-analytics-product.php [hunk 11]
```diff
@@ -77,6 +77,39 @@ public function get_id() {
...
+	/**
+	 * Returns the product variation id that should be reported to Analytics based on
+	 * product id admin settings.
+	 *
+	 * @return string|integer  The product variation id that should be reported to Analytics
+	 *
+	 * @since    1.7.0
+	 */
+	public function get_variable_id() {
+		$product_id = $this->get_custom_product_id($this->product);
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: We should move this line after the above check for "variable" products.
If it is a variable product and the `parent_id_enabled` setting is set to 'no' (aka "Send variation Unique ID") the custom product id should be derived from a variation product and not the parent, since there is a greater chance to match the correct product ID.

File: includes/class-wc-skroutz-analytics-product.php [hunk 12]
```diff
@@ -77,6 +77,39 @@ public function get_id() {
...
+
+		if ( $this->product->is_type( 'variable' ) && $this->items_product_id_settings['parent_id_enabled'] == 'no' ) {
+			$this->product = wc_get_product($this->product->get_children()[0]);
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: A product can be non visible (enabled property is not checked on admin panel). In order to increase our chances we should get the first visible product id:

```
// TODO Use only get_visible_children when we drop support for WooCommerce < 3.0
if ( method_exists( $this->product, 'get_visible_children' ) ) {
  $this->product = wc_get_product($this->product->get_visible_children()[0]);
} else {
  $this->product = wc_get_product($this->product->get_children(true)[0]);
}
```

File: includes/class-wc-skroutz-analytics-product.php [hunk 13]
```diff
@@ -77,6 +77,41 @@ public function get_id() {
...
+	 * for variable products.
+	 *
+	 * @return string|integer  The product id concatenated with first term id
+	 *
+	 * @since    1.7.0
+	 */
+	public function get_reviews_widget_product_id() {
+		if ( $this->product->is_type( 'variable' ) && $this->items_product_id_settings['parent_id_enabled'] == 'no' ) {
+			// TODO Use only get_visible_children when we drop support for WooCommerce < 3.0
+			$this->product = method_exists( $this->product, 'get_visible_children' ) ?
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: An error will occur if all the variation/children products are not visible. If that is the case we could fallback to the parent product. 
```php
// TODO Use only get_visible_children when we drop support for WooCommerce < 3.0
$variation_ids = method_exists( $this->product, 'get_visible_children' ) ?
  $this->product->get_visible_children() :
  $this->product->get_children(true);

// Verify there is at least one visible variation otherwise use the parent product
if ( isset($variation_ids[0]) ) {
  $this->product = wc_get_product($variation_ids[0]);
}
```

File: includes/class-wc-skroutz-analytics-product.php [hunk 14]
```diff
@@ -77,6 +77,41 @@ public function get_id() {
...
+
+		if ( ! $product_id ) {
+			$product_id = $this->items_product_id_settings['id'] == 'sku'
+				? $this->product->get_sku()
+				: $this->product->get_id();
+		}
+
+		if ( ! $product_id ) {
+			$product_id = "wc-sa-{$this->product->get_id()}";
+		}
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: We use the "wc-sa-" id when the sku field is not set in the product. I believe that is useless in the widget id and we could fallback to the product id.

```php
if ( ! $product_id && $this->items_product_id_settings['id'] == 'sku' ) {
  $product_id = $this->product->get_sku();
}

if ( ! $product_id ) {
  $product_id = $this->product->get_id();
}
```

File: includes/class-wc-skroutz-analytics-product.php [hunk 15]
```diff
@@ -148,4 +183,34 @@ private function get_product_terms_suffix() {
...
+	 */
+	private function get_product_first_term_suffix() {
+		$product_attributes = $this->product->get_variation_attributes();
+
+		$terms = null;
+		foreach ($this->items_product_id_settings['grouping_attributes'] as $attribute_name) {
+
+			$taxonomy_name = wc_attribute_taxonomy_name($attribute_name);
+
+			if ( isset($product_attributes[$taxonomy_name]) ) {
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: We should make sure the first element is available here:
```php
if ( isset($product_attributes[$taxonomy_name][0]) ) {
```

File: includes/class-wc-skroutz-analytics-product.php [hunk 16]
```diff
@@ -148,4 +183,34 @@ private function get_product_terms_suffix() {
 
 		return $terms;
 	}
+
+	/**
+	 * Get the suffix of the id that we send to sku reviews widget. The suffix contains for every variation
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-product.php]: I would replace `send` with `use`

Comment (harrisred) [reply includes/class-wc-skroutz-analytics-product.php]: Also I would rephrase "The suffix contains for every variation the first term id" to "The suffix is the first term of every grouped attribute in settings"

File: README.md [hunk 17]
```diff
@@ -26,6 +26,35 @@ The plugin is available from the [Wordpress plugin repo][11].
...
+    add_action( 'wc_skroutz_analytics_product_id_filter', 'my_product_id', 10, 2 );
+    ```
+
+2. The filter tag is: `wc_skroutz_analytics_product_reviews_widget_id`, the value is the generated
+   Unique ID with the term ID and the additional argument is the product object (WC_Product | WC_Product_Variable).
+
+   Usage example:
+
+    ```
+    function my_product_variation_id( $variation_id, $product ) {
```
Comment (harrisred) [inline README.md]: [minor] I would rename the function name as well to sth like: "my_product_reviews_widget_id". Don't forget the change the callback name in the add_action as well.

File: README.md [hunk 18]
```diff
@@ -132,5 +149,4 @@ Skroutz Analytics WooCommerce Plugin is licensed under the [GNU Public License v
 [9]: LICENSE.txt
 [10]: https://merchants.skroutz.gr/merchants/account/settings/analytics
 [11]: https://wordpress.org/plugins/skroutz-analytics-woocommerce/
-[12]: https://developer.wordpress.org/plugins/hooks/filters/
-
```
Comment (harrisred) [inline README.md:136]: It is a good common practice to leave an empty line at the end of each file

**Comment by harrisred** [comment]
PR merged
